zerver/lib/api.py

```diff
@@ -0,0 +1,334 @@
+        if param_name in IGNORED_PARAMS:
+            continue
+
+        param_default = parameter.default
+        # inspect._empty is the internal type used by inspect to indicate not
+        # specified defaults.
+        if param_default is inspect._empty:
+            param_default = NotSpecified
+
+        # defaulting a value to None automatically wraps the type annotation
+        # with Optional. We explicitly unwrap it for the case of Annotated,
+        # which otherwise causes undesired behaviors.
+        # This is fixed in Python 3.11: https://github.com/python/cpython/issues/90353
+        param_type = type_hints[param_name]
+        if param_default is None and is_optional(param_type):
+            optional_args = get_args(param_type)
+            inner_type = optional_args[0] if optional_args[1] is type(None) else optional_args[1]
+            if is_annotated(inner_type):
+                param_type = inner_type
+
+        metadata: Optional[Metadata] = None
+        if get_origin(param_type) is Annotated:
+            # The first type is the underlying type of the parameter, the rest
+            # are metadata attached to Annotated.
+            param_type, *type_args = get_args(param_type)
+            for type_arg in type_args:
+                if not isinstance(type_arg, Metadata):
+                    continue
+                assert metadata is None, "Metadata can only be defined once per parameter"
+                metadata = type_arg
+        # Set metadata to a default early to avoid addition None-checks.
+        if metadata is None:
+            metadata = Metadata()
+
+        # Metadata defines a validator making sure that argument_type_is_body is
+        # incompatible with whence.
+        if metadata.argument_type_is_body:
+            request_var_name = "request"
+        else:
+            request_var_name = metadata.whence if metadata.whence is not None else param_name
+
+        process_parameters.append(
+            FuncParam(
+                param_name=param_name,
+                param_type=param_type,
+                request_var_name=request_var_name,
+                default=param_default,
+                metadata=metadata,
+            )
+        )
+
+    return ViewFuncInfo(
+        view_func_full_name=view_func_full_name,
+        parameters=process_parameters,
+        no_nesting=True,
+    )
+
+
+def parse_request(
+    request: HttpRequest,
+    view_func_info: ViewFuncInfo,
+    args: Sequence[object],
+    kwargs: Mapping[str, object],
+) -> CallInfo:
+    processed_request_var_names = set()
+    kwargs_to_update = {}
+    errors = []
+    for parameter in view_func_info.parameters:
+        if parameter.metadata.path_only:
+            # For path_only parameters, they should already have been passed via
+            # the URL, so there's no need for us to do anything.
+            #
+            # TODO: Run validators for path_only parameters for NewType.
```

Comment:
Is this TODO for later or is it something you need to address before we merge?

Reply:
The TODO is the same one from request.py. We leave it for later.